From 802fccae2cef6c189d0e65339577feccc7403bdb Mon Sep 17 00:00:00 2001
From: Stuart Menefy <stuart.menefy@codasip.com>
Date: Tue, 23 Jan 2024 10:42:23 +0000
Subject: [PATCH 01/19] hw/sd/ssi-sd: Fix response to CMD58 (READ_OCR)

FSBL expects the response to CMD58 (READ_OCR) to have the MSB set to 0
(indicating that the card is not idle). However qemu was returning 1
(card is idle). Given that this works on real hardware, update qemu to
match, returning 0.

Signed-off-by: Stuart Menefy <stuart.menefy@codasip.com>
---
 hw/sd/ssi-sd.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/hw/sd/ssi-sd.c b/hw/sd/ssi-sd.c
index a6cc1ad6c8..0b40cf98e9 100644
--- a/hw/sd/ssi-sd.c
+++ b/hw/sd/ssi-sd.c
@@ -180,7 +180,20 @@ static uint32_t ssi_sd_transfer(SSIPeripheral *dev, uint32_t val)
                 /* CMD8/CMD58 returns R3/R7 response */
                 DPRINTF("Returned R3/R7\n");
                 s->arglen = 5;
-                s->response[0] = 1;
+                /*
+                 * SD card specification documents the MSB of R3/R7
+                 * as being the same as R1:
+                 *   It is one byte long, and the MSB is always set to zero.
+                 *   The other bits are error indications.
+                 * Bit 0 indicates:
+                 *   In idle state: The card is in idle state and running
+                 *   the initializing process.
+                 *
+                 * Originally this code set bit 0.
+                 * However FSBL expects a response of 0, which presumably it
+                 * gets from real hardware.
+                 */
+                s->response[0] = 0;
                 memcpy(&s->response[1], longresp, 4);
             } else if (s->arglen != 4) {
                 BADF("Unexpected response to cmd %d\n", s->cmd);
-- 
2.43.0

